pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp

```diff
@@ -462,18 +436,131 @@ class StateVectorKokkos final
+            PL_ABORT_IF(gate_matrix.empty(),
+                        std::string("Operation does not exist for ") + opName +
+                            std::string(" and no matrix provided."));
+            return applyNCMultiQubitOp(vector2view(gate_matrix),
+                                       controlled_wires, controlled_values,
+                                       wires, inverse);
+        }
+    }
+
+    /**
+     * @brief Apply a controlled-multi qubit operator to the state vector using
+     * a matrix
+     *
+     * @param matrix Kokkos gate matrix in the device space.
+     * @param controlled_wires Control wires.
+     * @param controlled_values Control values (true or false).
+     * @param wires Wires to apply gate to.
+     * @param inverse Indicates whether to use adjoint of gate. (Default to
+     * false)
+     */
+    void applyNCMultiQubitOp(const KokkosVector matrix,
+                             const std::vector<std::size_t> &controlled_wires,
+                             const std::vector<bool> &controlled_values,
+                             const std::vector<std::size_t> &wires,
+                             bool inverse = false) {
+        const std::size_t num_qubits = this->getNumQubits();
+        const std::size_t two2N =
+            exp2(num_qubits - wires.size() - controlled_wires.size());
+        const std::size_t dim = exp2(wires.size());
+        KokkosVector matrix_trans("matrix_trans", matrix.size());
+
+        if (inverse) {
+            Kokkos::MDRangePolicy<DoubleLoopRank> policy_2d({0, 0}, {dim, dim});
+            Kokkos::parallel_for(
+                policy_2d, KOKKOS_LAMBDA(std::size_t i, std::size_t j) {
+                    matrix_trans(i + j * dim) =
+                        Kokkos::conj(matrix(i * dim + j));
+                });
+        } else {
+            matrix_trans = matrix;
+        }
+
+        switch (wires.size()) {
+        case 1:
+            Kokkos::parallel_for(two2N, applyNC1QubitOpFunctor<fp_t>(
+                                            *data_, num_qubits, matrix_trans,
+                                            controlled_wires, controlled_values,
+                                            wires));
+            break;
+        case 2:
+            Kokkos::parallel_for(two2N, applyNC2QubitOpFunctor<fp_t>(
+                                            *data_, num_qubits, matrix_trans,
+                                            controlled_wires, controlled_values,
+                                            wires));
+            break;
+        case 3:
+            Kokkos::parallel_for(two2N, applyNC3QubitOpFunctor<fp_t>(
+                                            *data_, num_qubits, matrix_trans,
+                                            controlled_wires, controlled_values,
+                                            wires));
+            break;
+        default:
+            // TODO: explore runtime determine scratch space level (L0 vs L1)
```

Comment:
I reckon we don't need L0 (register) scratch space for wires >= 3 due to capacity issues ;)

Reply:
Could be a good case for exploring https://kokkos.org/kokkos-core-wiki/API/simd/simd.html to a more general implementation than the AVX2/AVX512 kernels in LQ

I believe Christian Trott mentioned this would be an interesting thing in the past when we were in communication.